File: sdk/typescript/src/namespaces/email-debate.ts
/**
 * Email Debate Namespace API
 *
 * Provides endpoints for multi-agent email prioritization and triage.
 */

import type { AragoraClient } from '../client';

/**
 * Email input for prioritization
 */
export interface EmailInput {
  subject?: string;
  body?: string;
  sender: string;
  received_at?: string;
  message_id?: string;
  user_id?: string;
  thread_id?: string;
  labels?: string[];
  attachments?: Array<{
    filename: string;
    mime_type: string;
    size_bytes: number;
  }>;
}

/**
 * Priority level for emails
 */
export type PriorityLevel = 'critical' | 'high' | 'medium' | 'low' | 'none';

/**
 * Email prioritization result
 */
export interface PrioritizationResult {
  message_id?: string;
  priority: PriorityLevel;
  score: number;
  confidence: number;
  reasoning: string;
  suggested_actions: string[];
  category?: string;
  response_urgency?: 'immediate' | 'same_day' | 'this_week' | 'when_possible';
  agents_consulted: string[];
  consensus_reached: boolean;
  debate_id?: string;
}

/**
 * Batch prioritization result
 */
export interface BatchPrioritizationResult {
  results: PrioritizationResult[];
  total_processed: number;
  processing_time_ms: number;
  summary: {
    critical: number;
    high: number;
    medium: number;
    low: number;
    none: number;
  };
}

/**
 * Email category for triage
 */
export type EmailCategory =
  | 'action_required'
  | 'fyi'
  | 'meeting'
  | 'follow_up'
  | 'newsletter'
  | 'spam'
  | 'personal'
  | 'finance'
  | 'legal'
  | 'other';

/**
 * Triage result for an email
 */
export interface TriageResult {
  message_id?: string;
  category: EmailCategory;
  priority: PriorityLevel;
  suggested_folder?: string;
  suggested_labels?: string[];
  auto_reply_suggested: boolean;
  auto_reply_draft?: string;
  delegate_to?: string;
  snooze_until?: string;
  reasoning: string;
}

/**
 * Inbox triage response
 */
export interface InboxTriageResponse {
  results: TriageResult[];
  total_triaged: number;
  processing_time_ms: number;
  summary: {
    by_category: Record<EmailCategory, number>;
    by_priority: Record<PriorityLevel, number>;
    auto_replies_suggested: number;
    delegations_suggested: number;
  };
}

/**
 * Email Debate namespace for AI-powered email prioritization.
 *
 * @example
 * ```typescript
 * // Prioritize a single email
 * const result = await client.emailDebate.prioritize({
 *   subject: 'Q4 Budget Review - Action Required',
 *   body: 'Please review the attached budget proposal...',
 *   sender: 'cfo@company.com'
 * });
 * console.log(`Priority: ${result.priority}, Score: ${result.score}`);
 *
 * // Batch prioritize inbox
 * const batch = await client.emailDebate.prioritizeBatch([
 *   { subject: 'Meeting tomorrow', sender: 'bob@example.com' },
 *   { subject: 'Invoice #1234', sender: 'billing@vendor.com' }
 * ]);
 * console.log(`Processed ${batch.total_processed} emails`);
 *
 * // Full inbox triage
 * const triage = await client.emailDebate.triageInbox(emails);
 * triage.results.forEach(r => {
 *   console.log(`${r.category}: ${r.priority} - ${r.suggested_folder}`);
 * });
 * ```
 */
export class EmailDebateNamespace {
  constructor(private client: AragoraClient) {}

  /**
   * Prioritize a single email using multi-agent debate.
   *
   * Agents analyze the email content, sender reputation, and context
   * to determine priority and suggest actions.
   *
   * @param email - Email to prioritize
   * @param options.userId - User ID for personalized prioritization
   */
  async prioritize(
    email: EmailInput,
    options?: { userId?: string }
  ): Promise<PrioritizationResult> {
    const body = { ...email };
    if (options?.userId) {
      body.user_id = options.userId;
    }
    return this.client.request<PrioritizationResult>('POST', '/api/v1/email/prioritize', body);
  }

  /**
   * Prioritize multiple emails in batch.
   *
   * More efficient than calling prioritize() multiple times.
   *
   * @param emails - Array of emails to prioritize
   * @param options.userId - User ID for personalized prioritization
   * @param options.parallel - Process emails in parallel (default: true)
   */
  async prioritizeBatch(
    emails: EmailInput[],
    options?: { userId?: string; parallel?: boolean }
  ): Promise<BatchPrioritizationResult> {
    const body = {
      emails,
      user_id: options?.userId,
      parallel: options?.parallel ?? true,
    };
    return this.client.request<BatchPrioritizationResult>(
      'POST',
      '/api/v1/email/prioritize/batch',
      body
    );
  }

  /**
   * Triage inbox with full categorization.
   *
   * Provides comprehensive email triage including:
   * - Category assignment (action_required, fyi, meeting, etc.)
   * - Priority scoring
   * - Folder suggestions
   * - Auto-reply drafts
   * - Delegation recommendations
   *
   * @param emails - Emails to triage
   * @param options.userId - User ID for personalized triage
   * @param options.includeAutoReplies - Generate auto-reply drafts
   */
  async triageInbox(
    emails: EmailInput[],
    options?: { userId?: string; includeAutoReplies?: boolean }
  ): Promise<InboxTriageResponse> {
    const body = {
      emails,
      user_id: options?.userId,
      include_auto_replies: options?.includeAutoReplies ?? false,
    };
    return this.client.request<InboxTriageResponse>('POST', '/api/v1/email/triage', body);
  }

  /**
   * Get prioritization history for a user.
   *
   * @param options.userId - User ID
   * @param options.limit - Maximum results (default: 50)
   * @param options.since - Filter to emails since this timestamp
   */
  async getHistory(options: {
    userId: string;
    limit?: number;
    since?: string;
  }): Promise<PrioritizationResult[]> {
    const params = new URLSearchParams();
    params.set('user_id', options.userId);
    if (options.limit) params.set('limit', options.limit.toString());
    if (options.since) params.set('since', options.since);

    return this.client.request<PrioritizationResult[]>(
      'GET',
      `/api/v1/email/prioritize/history?${params.toString()}`
    );
  }
}

File: sdk/python/aragora/namespaces/email_debate.py
"""
Email Debate Namespace API

Provides methods for multi-agent email prioritization and triage:
- Single email prioritization
- Batch email prioritization
- Full inbox triage with categorization
- Prioritization history retrieval
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class EmailDebateAPI:
    """
    Synchronous Email Debate API.

    Provides AI-powered email prioritization using multi-agent debate:
    - Prioritize single emails with reasoning and suggested actions
    - Batch prioritize multiple emails efficiently
    - Full inbox triage with categorization and auto-reply drafts
    - Retrieve prioritization history

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> result = client.email_debate.prioritize(
        ...     email={
        ...         "subject": "Q4 Budget Review - Action Required",
        ...         "body": "Please review the attached budget proposal...",
        ...         "sender": "cfo@company.com",
        ...     },
        ...     user_id="user_123",
        ... )
        >>> print(f"Priority: {result['priority']}, Score: {result['score']}")
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # ===========================================================================
    # Email Prioritization
    # ===========================================================================

    def prioritize(
        self,
        email: dict[str, Any],
        user_id: str | None = None,
    ) -> dict[str, Any]:
        """
        Prioritize a single email using multi-agent debate.

        Agents analyze the email content, sender reputation, and context
        to determine priority and suggest actions.

        Args:
            email: Email to prioritize. Should include:
                - subject: Email subject line (optional)
                - body: Email body content (optional)
                - sender: Sender email address (required)
                - received_at: ISO timestamp when received (optional)
                - message_id: Unique message identifier (optional)
                - user_id: User ID for the email owner (optional)
                - thread_id: Thread/conversation ID (optional)
                - labels: List of labels/tags (optional)
                - attachments: List of attachment dicts with filename,
                  mime_type, size_bytes (optional)
            user_id: User ID for personalized prioritization

        Returns:
            Dict with prioritization result including:
                - priority: Priority level (critical/high/medium/low/none)
                - score: Numeric priority score
                - confidence: Confidence in the prioritization
                - reasoning: Explanation for the priority assignment
                - suggested_actions: List of recommended actions
                - category: Email category (optional)
                - response_urgency: When to respond (optional)
                - agents_consulted: List of agents that participated
                - consensus_reached: Whether agents reached consensus
                - debate_id: ID of the debate session (optional)
        """
        data: dict[str, Any] = {**email}
        if user_id is not None:
            data["user_id"] = user_id
        return self._client.request("POST", "/api/v1/email/prioritize", json=data)

    def prioritize_batch(
        self,
        emails: list[dict[str, Any]],
        user_id: str | None = None,
        parallel: bool = True,
    ) -> dict[str, Any]:
        """
        Prioritize multiple emails in batch.

        More efficient than calling prioritize() multiple times.

        Args:
            emails: List of email dicts to prioritize. Each email should
                include subject, body, sender, and optionally received_at,
                message_id, user_id, thread_id, labels, and attachments.
            user_id: User ID for personalized prioritization
            parallel: Process emails in parallel (default: True)

        Returns:
            Dict with batch prioritization results including:
                - results: List of prioritization results for each email
                - total_processed: Number of emails processed
                - processing_time_ms: Total processing time in milliseconds
                - summary: Counts by priority level (critical/high/medium/low/none)
        """
        data: dict[str, Any] = {
            "emails": emails,
            "parallel": parallel,
        }
        if user_id is not None:
            data["user_id"] = user_id
        return self._client.request("POST", "/api/v1/email/prioritize/batch", json=data)

    # ===========================================================================
    # Inbox Triage
    # ===========================================================================

    def triage_inbox(
        self,
        emails: list[dict[str, Any]],
        user_id: str | None = None,
        include_auto_replies: bool = False,
    ) -> dict[str, Any]:
        """
        Triage inbox with full categorization.

        Provides comprehensive email triage including:
        - Category assignment (action_required, fyi, meeting, follow_up,
          newsletter, spam, personal, finance, legal, other)
        - Priority scoring
        - Folder suggestions
        - Auto-reply drafts (if enabled)
        - Delegation recommendations

        Args:
            emails: List of email dicts to triage. Each email should include
                subject, body, sender, and optionally received_at, message_id,
                user_id, thread_id, labels, and attachments.
            user_id: User ID for personalized triage
            include_auto_replies: Generate auto-reply drafts (default: False)

        Returns:
            Dict with triage results including:
                - results: List of triage results for each email with category,
                  priority, suggested_folder, suggested_labels, auto_reply_suggested,
                  auto_reply_draft, delegate_to, snooze_until, and reasoning
                - total_triaged: Number of emails triaged
                - processing_time_ms: Total processing time in milliseconds
                - summary: Aggregated stats by category and priority
        """
        data: dict[str, Any] = {
            "emails": emails,
            "include_auto_replies": include_auto_replies,
        }
        if user_id is not None:
            data["user_id"] = user_id
        return self._client.request("POST", "/api/v1/email/triage", json=data)

    # ===========================================================================
    # History
    # ===========================================================================

    def get_history(
        self,
        user_id: str,
        limit: int | None = None,
        since: str | None = None,
    ) -> dict[str, Any]:
        """
        Get prioritization history for a user.

        Args:
            user_id: User ID to get history for
            limit: Maximum number of results (default: 50)
            since: Filter to emails since this ISO timestamp

        Returns:
            Dict with list of past prioritization results
        """
        params: dict[str, Any] = {"user_id": user_id}
        if limit is not None:
            params["limit"] = limit
        if since is not None:
            params["since"] = since
        return self._client.request("GET", "/api/v1/email/prioritize/history", params=params)


class AsyncEmailDebateAPI:
    """
    Asynchronous Email Debate API.

    Provides AI-powered email prioritization using multi-agent debate.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     result = await client.email_debate.prioritize(
        ...         email={
        ...             "subject": "Urgent: Server Down",
        ...             "body": "Production server is unresponsive...",
        ...             "sender": "alerts@monitoring.com",
        ...         },
        ...         user_id="user_123",
        ...     )
        ...     print(f"Priority: {result['priority']}")
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # ===========================================================================
    # Email Prioritization
    # ===========================================================================

    async def prioritize(
        self,
        email: dict[str, Any],
        user_id: str | None = None,
    ) -> dict[str, Any]:
        """
        Prioritize a single email using multi-agent debate.

        Args:
            email: Email to prioritize with subject, body, sender, etc.
            user_id: User ID for personalized prioritization

        Returns:
            Dict with prioritization result including priority, score,
            confidence, reasoning, and suggested actions
        """
        data: dict[str, Any] = {**email}
        if user_id is not None:
            data["user_id"] = user_id
        return await self._client.request("POST", "/api/v1/email/prioritize", json=data)

    async def prioritize_batch(
        self,
        emails: list[dict[str, Any]],
        user_id: str | None = None,
        parallel: bool = True,
    ) -> dict[str, Any]:
        """
        Prioritize multiple emails in batch.

        Args:
            emails: List of email dicts to prioritize
            user_id: User ID for personalized prioritization
            parallel: Process emails in parallel (default: True)

        Returns:
            Dict with batch results, total_processed, and summary by priority
        """
        data: dict[str, Any] = {
            "emails": emails,
            "parallel": parallel,
        }
        if user_id is not None:
            data["user_id"] = user_id
        return await self._client.request("POST", "/api/v1/email/prioritize/batch", json=data)

    # ===========================================================================
    # Inbox Triage
    # ===========================================================================

    async def triage_inbox(
        self,
        emails: list[dict[str, Any]],
        user_id: str | None = None,
        include_auto_replies: bool = False,
    ) -> dict[str, Any]:
        """
        Triage inbox with full categorization.

        Args:
            emails: List of email dicts to triage
            user_id: User ID for personalized triage
            include_auto_replies: Generate auto-reply drafts (default: False)

        Returns:
            Dict with triage results including category, priority,
            suggested_folder, and optional auto-reply drafts
        """
        data: dict[str, Any] = {
            "emails": emails,
            "include_auto_replies": include_auto_replies,
        }
        if user_id is not None:
            data["user_id"] = user_id
        return await self._client.request("POST", "/api/v1/email/triage", json=data)

    # ===========================================================================
    # History
    # ===========================================================================

    async def get_history(
        self,
        user_id: str,
        limit: int | None = None,
        since: str | None = None,
    ) -> dict[str, Any]:
        """
        Get prioritization history for a user.

        Args:
            user_id: User ID to get history for
            limit: Maximum number of results (default: 50)
            since: Filter to emails since this ISO timestamp

        Returns:
            Dict with list of past prioritization results
        """
        params: dict[str, Any] = {"user_id": user_id}
        if limit is not None:
            params["limit"] = limit
        if since is not None:
            params["since"] = since
        return await self._client.request("GET", "/api/v1/email/prioritize/history", params=params)
